Update SDUI technotes (#574)
apollographql/docs: src/content/technotes/TN0032-sdui-basics.mdx

## Schema design

Take, for example, this mock e-commerce schema, which allows clients to display a list of featured products:

```graphql
type Product @key(fields: "id") {
  id: ID!
  name: String
  formattedPrice: String
  # other product fields...
}

type ProductsCarousel {
  products: [Product]
  count: Int
}

type ProductsList {
  products: [Product]
  count: Int
}

type ProductsError {
  message: String!
}

union FeaturedProductsResponse = ProductsList | ProductsCarousel | ProductsError

type Query {
  featuredProducts: FeaturedProductsResponse
}
```

While querying `Query.featuredProducts`, the client can render multiple specific experiences depending on the query result. You can use the returned `__typename` value to determine whether to render a collection of products or show the user an error message.

```graphql
fragment productFields on Product {
  id
  name
  formattedPrice
}

query FeaturedProductsCollection {
  featuredProducts {
    __typename
    ... on ProductsList {
      products {
        ...productFields
      }
    }
    ... on ProductsCarousel {
      products {
        ...productFields
      }
    }
    ... on ProductsError {
      message
    }
  }

> }

Reviewer: ```suggestion
}

```

Check out [SDUI schema design](TN0033-sdui-schema-design/) for more design patterns.

```

---
title: Server-driven UI basics
id: TN0032
tags: [schema-design, server-driven-ui]
---

> Server-driven UI (SDUI) is an architectural pattern that aims to reduce client-side logic and provide consistency across client platforms (web, iOS,Android, etc.) by returning product information from an API instead of domain data. This approach is geared toward use cases where collaboration between front-end and back-end teams is possible, and there is a strong desire to build a cohesive experience across multiple teams and organizations.

Reviewer: ```suggestion
Server-driven UI (SDUI) is an architectural pattern that aims to reduce client-side logic and provide consistency across client platforms (web, iOS, Android, etc.) by returning product information from an API instead of domain data. This approach is geared toward use cases where collaboration between front-end and back-end teams is possible, and there is a strong desire to build a cohesive experience across multiple teams and organizations.
```

> For advanced patterns and schema types, see [more SDUI topics](/technotes/tags/server-driven-ui).

## Architectural goals

> The primary need for SDUI comes from mobile native-app development. These platforms require shipping versions of your client-side app to a store that usually has a lengthy review process before a new version gets published. Then, even after a new app version is released, users have to update their versions from the store to start using or experiencing the new features.

Reviewer: ```suggestion
The primary need for SDUI comes from mobile native-app development. These platforms require shipping versions of your client-side app to an app store that usually has a lengthy review process before a new version gets published. Then, even after a new app version is released, users must update their versions to start using or experiencing the new features.
```

If your API returns exactly what to display on the screen, you can control the different UI elements or text without app version updates. You can also change the inputs sent to the API without requiring any app updates.

When building a graph with server-driven UI, converging the schema with the design language used by client teams is often desirable. This requires a definitive design system and a well-defined collection of components to render views on the frontend.

## Guiding principles

> SDUI is not a framework or library you can install. It's a design pattern you can implement in different ways while still adhering to the same core principles. You may start with just a few parts of your app controlled by the server, like text or whether to hide certain UI elements. SDUI could then advance to more parts of your app, like changing theming or component features from the API, all the way to having a generic layout that can render any UI component in any order.

Reviewer: ```suggestion
SDUI is not a framework or library you can install. It's a design pattern you can implement in different ways while still adhering to the same core principles.
```

### Start with UI designs

> Start with your UI designs instead of looking at data models or API responses.  Take a [demand-driven schema design approach](/technotes/TN0027-demand-oriented-schema-design/) write your preferred query from the client's perspective. The demand-driven concept can apply to any client, but UI applications should **have a direct connection from the API response to items on the screen**.

Reviewer: ```suggestion
Start with your UI designs instead of looking at data models or API responses.  Take a [demand-driven schema design approach](/technotes/TN0027-demand-oriented-schema-design/) to write your preferred query from the client's perspective. The demand-driven concept can apply to any client, but UI applications should **have a direct connection from the API response to items on the screen**.
```

If you find yourself describing steps to go from API to views, that's a warning sign that you'll have to duplicate that logic across clients.

### Reduce logic in existing client code

> Next, identify any logic in your existing client code. Some may think of "logic" as only business rules in the backend. However, if you've ever had to update conditionals or switch statements for a new feature, or to update how a feature works, you can consider these frontend portions as "logic," too. Any other places that accept your API response and then transform, concatenate, or use helper functions are additional warning signs for logic. 

Reviewer: ```suggestion
Identify and reduce any logic in your existing client code. Some may think of "logic" as only business rules in the backend. However, if you've ever had to update conditionals or switch statements for a new feature, or to update how a feature works, you can consider these frontend portions as "logic," too. Any other places that accept your API response and then transform, concatenate, or use helper functions are additional warning signs for logic. 
```

## Schema design

Take, for example, this mock e-commerce schema, which allows clients to display a list of featured products:

```graphql
type Product @key(fields: "id") {
  id: ID!
  name: String
  formattedPrice: String
  # other product fields...
}

type ProductsCarousel {
  products: [Product]
  count: Int
}

type ProductsList {
  products: [Product]
  count: Int
}

type ProductsError {
  message: String!
}

union FeaturedProductsResponse = ProductsList | ProductsCarousel | ProductsError

type Query {
  featuredProducts: FeaturedProductsResponse
}
```

While querying `Query.featuredProducts`, the client can render multiple specific experiences depending on the query result. You can use the returned `__typename` value to determine whether to render a collection of products or show the user an error message.

```graphql
fragment productFields on Product {
  id
  name
  formattedPrice
}

query FeaturedProductsCollection {
  featuredProducts {
    __typename
    ... on ProductsList {
      products {
        ...productFields
      }
    }
    ... on ProductsCarousel {
      products {
        ...productFields
      }
    }
    ... on ProductsError {
      message
    }
  }
}

> ```

Reviewer: ```suggestion
```

Check out [SDUI schema design](TN0033-sdui-schema-design/) for more design patterns.
```
